ledger/complete/wal/wal.go

```diff
@@ -146,33 +151,50 @@ func (w *LedgerWAL) replay(
 	}
 
 	if useCheckpoints {
-		latestCheckpoint, err := checkpointer.LatestCheckpoint()
+		allCheckpoints, err := checkpointer.ListCheckpoints()
 		if err != nil {
-			return fmt.Errorf("cannot get latest checkpoint: %w", err)
+			return fmt.Errorf("cannot get list of checkpoints: %w", err)
 		}
 
-		if latestCheckpoint != -1 && latestCheckpoint+1 >= from { //+1 to account for connected checkpoint and segments
+		var availableCheckpoints []int
+
+		// if there are no checkpoints already, don't bother
+		if len(allCheckpoints) > 0 {
+			// from-1 to account for checkpoints connected to segments, ie. checkpoint 8 if replaying segments 9-12
+			availableCheckpoints = getPossibleCheckpoints(allCheckpoints, from-1, to)
+		}
+
+		for len(availableCheckpoints) > 0 { // as long as there are checkpoints to try
+			latestCheckpoint := availableCheckpoints[len(availableCheckpoints)-1]
```

Comment:
```suggestion
		for len(availableCheckpoints) > 0 {
			// as long as there are checkpoints to try, we alway try with the last checkpoint file, since 
			// it allows us to load less segments.
			latestCheckpoint := availableCheckpoints[len(availableCheckpoints)-1]
```